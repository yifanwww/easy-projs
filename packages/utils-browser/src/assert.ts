export class AssertionError extends Error {}

/**
 * @param value The input that is checked for being truthy.
 */
export function assert(value: unknown, message?: string): asserts value {
    if (!value) {
        throw new AssertionError(message ?? 'Assertion Error.');
    }
}

export function assertIsString(value: unknown, varName: string): asserts value is string {
    assert(typeof value === 'string', `${varName} should be string.`);
}

export function assertIsNumber(value: unknown, varName: string): asserts value is number {
    assert(typeof value === 'number', `${varName} should be number.`);
}

export function assertIsBoolean(value: unknown, varName: string): asserts value is boolean {
    assert(typeof value === 'boolean', `${varName} should be boolean.`);
}

/**
 * The function for asserting whether a value's type is `never`.
 */
export function assertIsNever(value: never): never {
    throw new AssertionError(`"${String(value)}" should be "never" type.`);
}

/**
 * The function for asserting whether a value is defined (!== `undefined` && !== `null`).
 */
export function assertIsDefined<T>(value: T, varName = 'value'): asserts value is NonNullable<T> {
    if (value === undefined || value === null) {
        throw new AssertionError(`Expected '${varName}' to be defined, but received \`${value as null | undefined}\``);
    }
}
