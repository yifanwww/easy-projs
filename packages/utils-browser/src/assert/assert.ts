export class AssertionError extends Error {}

/**
 * @param value The input that is checked for being truthy.
 */
export function assert(value: unknown, message?: string): asserts value {
    if (!value) {
        throw new AssertionError(message ?? 'Assertion Error.');
    }
}

export function assertIsString(value: unknown, varName: string): asserts value is string {
    assert(typeof value === 'string', `${varName} should be string.`);
}

export function assertIsNumber(value: unknown, varName: string): asserts value is number {
    assert(typeof value === 'number', `${varName} should be number.`);
}

export function assertIsBoolean(value: unknown, varName: string): asserts value is boolean {
    assert(typeof value === 'boolean', `${varName} should be boolean.`);
}

/**
 * The function for asserting whether a value's type is `never`.
 */
export function assertIsNever(value: never): never {
    throw new AssertionError(`"${String(value)}" should be "never" type.`);
}
