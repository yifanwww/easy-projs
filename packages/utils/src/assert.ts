/**
 * @param value The input that is checked for being truthy.
 */
export function assert(value: unknown): asserts value {
    if (!value) {
        throw new Error('Assertion Error');
    }
}

export function assertIsString(value: unknown, varName: string): asserts value is string {
    if (typeof value !== 'string') {
        throw new Error(`Assertion Error: ${varName} should be string`);
    }
}

export function assertIsNumber(value: unknown, varName: string): asserts value is number {
    if (typeof value !== 'number') {
        throw new Error(`Assertion Error: ${varName} should be number`);
    }
}

export function assertIsBoolean(value: unknown, varName: string): asserts value is boolean {
    if (typeof value !== 'boolean') {
        throw new Error(`Assertion Error: ${varName} should be boolean`);
    }
}
