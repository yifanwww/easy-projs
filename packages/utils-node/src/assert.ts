import { AssertionError } from 'node:assert';

/**
 * The function for asserting whether a value's type is `never`.
 */
export function assertIsNever(value: never): never {
    throw new AssertionError({
        message: `"${String(value)}" should be "never" type.`,
        actual: value,
    });
}

/**
 * The function for asserting whether a value is defined (!== `undefined` && !== `null`).
 */
export function assertIsDefined<T>(value: T, varName = 'value'): asserts value is NonNullable<T> {
    if (value === undefined || value === null) {
        throw new AssertionError({
            message: `Expected '${varName}' to be defined, but received \`${value as null | undefined}\``,
            actual: value,
        });
    }
}
