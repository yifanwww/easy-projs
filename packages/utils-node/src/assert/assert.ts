import { AssertionError } from 'node:assert';

/**
 * The function for asserting whether a value's type is `never`.
 */
export function assertIsNever(value: never): never {
    throw new AssertionError({ message: `"${String(value)}" should be "never" type.` });
}
