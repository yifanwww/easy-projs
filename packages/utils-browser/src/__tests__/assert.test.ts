import { describe, expect, it } from '@jest/globals';

import { AssertionError, assert, assertIsBoolean, assertIsNever, assertIsNumber, assertIsString } from '../assert.js';

describe(`Test fn \`${assert.name}\``, () => {
    it('should assert an expression', () => {
        const str = 'hello world';

        expect(() => assert(typeof str === 'string')).not.toThrow();
        expect(() => assert(typeof str === 'number')).toThrow(AssertionError);
        expect(() => assert(typeof str === 'number')).toThrowErrorMatchingSnapshot();
    });
});

describe(`Test fn \`${assertIsBoolean.name}\``, () => {
    it('should assert a value is boolean', () => {
        const bool = true;
        const num = 0;

        expect(() => assertIsBoolean(bool, 'bool')).not.toThrow();
        expect(() => assertIsBoolean(num, 'num')).toThrow(AssertionError);
        expect(() => assertIsBoolean(num, 'num')).toThrowErrorMatchingSnapshot();
    });
});

describe(`Test fn \`${assertIsNumber.name}\``, () => {
    it('should assert a value is number', () => {
        const num = 0;
        const str = 'hello world';

        expect(() => assertIsNumber(num, 'num')).not.toThrow();
        expect(() => assertIsNumber(str, 'str')).toThrow(AssertionError);
        expect(() => assertIsNumber(str, 'str')).toThrowErrorMatchingSnapshot();
    });
});

describe(`Test fn \`${assertIsString.name}\``, () => {
    it('should assert a value is string', () => {
        const str = 'hello world';
        const num = 0;

        expect(() => assertIsString(str, 'str')).not.toThrow();
        expect(() => assertIsString(num, 'num')).toThrow(AssertionError);
        expect(() => assertIsString(num, 'num')).toThrowErrorMatchingSnapshot();
    });
});

describe(`Test fn \`${assertIsNever.name}\``, () => {
    it('should throw an error', () => {
        expect(() => assertIsNever('hello world' as never)).toThrow(AssertionError);
        expect(() => assertIsNever('hello world' as never)).toThrowErrorMatchingSnapshot();
    });
});
