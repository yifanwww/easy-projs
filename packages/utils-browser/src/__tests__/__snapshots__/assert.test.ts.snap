// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test fn \`assert\` should assert an expression 1`] = `"Assertion Error."`;

exports[`Test fn \`assertIsBoolean\` should assert a value is boolean 1`] = `"num should be boolean."`;

exports[`Test fn \`assertIsDefined\` should throw an error if assertion failes 1`] = `"Expected 'value' to be defined, but received \`null\`"`;

exports[`Test fn \`assertIsDefined\` should throw an error if assertion failes 2`] = `"Expected 'variable' to be defined, but received \`undefined\`"`;

exports[`Test fn \`assertIsNever\` should throw an error 1`] = `""hello world" should be "never" type."`;

exports[`Test fn \`assertIsNumber\` should assert a value is number 1`] = `"str should be number."`;

exports[`Test fn \`assertIsString\` should assert a value is string 1`] = `"num should be string."`;
